### ISSUE TITLE
Avoid hashing more than is strictly necessary, in the compiler.

### ISSUE THREAD

**Issue by eddyb** [rust-lang/rust#56308]
It should be possible to build cheaper interners than what we currently have, using `raw_entry`.
cc @rust-lang/compiler @Gankro @Amanieu

**Comment by Zoxc** [comment]
I have [a branch](https://github.com/rust-lang/rust/compare/master...Zoxc:int-ext?expand=1) which adds functions for interning based on [this PR](https://github.com/rust-lang/rust/pull/50959). I suggest that further work be based on that.

cc @nnethercote

Also relevant https://github.com/rust-lang/rust/issues/55514 https://github.com/rust-lang/rust/pull/56241

**Comment by Zoxc** [comment]
Sure. I think we have to change things to use a `HashMap` anyway. `HashSet` hides all the fun bits.

**Comment by eddyb** [comment]
Oh, another place where we could take advantage of computing the hash only once is the query engine, in case we had to compute a query, and we're inserting the result.

**Comment by michaelwoerister** [comment]
Another note, not directly related to interners: We have some things in the compiler that already contain high-quality hash values of themselves. E.g. it's unnecessary work to hash a `DefPathHash`. There's at least one such map here: https://github.com/rust-lang/rust/blob/691a7f8e2b2de984c3e0f062cbd294763af3d09a/src/librustc/ty/context.rs#L919

There are probably a few additional, similar cases.

**Comment by eddyb** [comment]
@michaelwoerister I think we can maybe use a wrapper for the hash state that uses specialization to avoid double-hashing, in that case, instead of relying on `raw_entry`.

**Comment by michaelwoerister** [comment]
I think the raw_entry API would be the nicer way to do it. It seems to be tailored to support a case just like this (among other use cases).

**Comment by eddyb** [comment]
@michaelwoerister Yes but it would require making an entire wrapper of the `HashMap` public API which carefully avoids letting anything hash the key (going as far as wrapping the key in a type that panics in its `Hash` `impl`).

**Comment by eddyb** [comment]
I'll reopen this because the discussion mentioned things other than interners.

Also, I still have to write those code examples for @michaelwoerister.

**Comment by Nilstrieb** [comment]
The interners currently do use the raw entry APIs: https://github.com/rust-lang/rust/blob/cba14074bb4cc12bfe918eabd0d52a3999b2a461/compiler/rustc_data_structures/src/sharded.rs#L99

**Comment by Mark-Simulacrum** [comment]
I ran some analysis to find the maps changed in #120076 by looking for things put into HashMaps that looked like they were already hashes, using a few heuristics (e.g., similar number of 1 and 0 bits). I think that found most or all of the problem places. I don't see a good way to assert we don't regress there -- with a TypeId for non 'static things, we could assert in Hash64/Hash128 etc impls that we're not hashing them with FxHash (or that we're only hashing them with Unhash). But that assert doesn't exist today, and I don't think adding e.g. specialization or a doc(hidden) method on Hasher in std makes sense for this at this time.

I think the next big opportunity for reducing hashing would be maps with dense, relatively small integer valued keys, where we actually don't need to hash because they're sufficiently dense that a vec makes more sense. We saw big wins with https://github.com/rust-lang/rust/pull/119977 from both (a) removing the hashing and (b) improved cache locality due to keeping things closer together. But those typically aren't as easy of a transformation to make, and to some extent are a time/space tradeoff.
